lambda-buffers-compiler/src/LambdaBuffers/Compiler/KindCheck.hs

```diff
@@ -118,210 +101,123 @@ globalStrategy = reinterpret $ \case
-          Left l -> PC.UnboundTyVarError td l modName
-      InferUnifyTermErr (I.Constraint (k1, k2)) ->
-        throwError . PC.CompKindCheckError $ PC.IncorrectApplicationError td (kind2ProtoKind k1) (kind2ProtoKind k2) modName
-      InferRecursiveSubstitutionErr _ ->
-        throwError . PC.CompKindCheckError $ PC.RecursiveKindError td modName
-      InferImpossibleErr t ->
-        throwError . PC.InternalError $ t
-
-    -- Gets the original term associated with the Variable.
-    getTermType :: PC.ModuleName -> PC.TyDef -> Variable -> Eff effs (Either PC.TyVar PC.TyRef)
-    getTermType modName td va = do
-      let termMap = snd . run . runWriter . runReader modName $ tyDef2Map td
-      case termMap M.!? va of
-        Just x -> pure x
-        Nothing ->
-          throwError . PC.InternalError . pack . unlines $
-            [ "Could not find the corresponding source info for:"
-            , show . pretty $ va
-            , "This should never happen."
-            , "Please report error."
-            ]
-      where
-        {- Conversion functions that associate a variable with its KC term.
-        The Monad has a Writer Map instance which is then used to retrieve sourceInfo
-        about the term.
-        -}
-
-        tyDef2Map :: PC.TyDef -> HandleErrorEnv ()
-        tyDef2Map = tyAbs2Map . view #tyAbs
-
-        -- NOTE(cstml): Is there any purpose for anything from the tyArg - be sure to cover with tests.
-        tyAbs2Map :: PC.TyAbs -> HandleErrorEnv ()
-        tyAbs2Map tyAbs = tyBody2Map (tyAbs ^. #tyBody)
-
-        tyBody2Map :: PC.TyBody -> HandleErrorEnv ()
-        tyBody2Map = \case
-          PC.OpaqueI _ -> pure ()
-          PC.SumI s -> sum2Map s
-
-        sum2Map :: PC.Sum -> HandleErrorEnv ()
-        sum2Map (PC.Sum constr _) = traverse_ constr2Map $ M.elems constr
-
-        constr2Map :: PC.Constructor -> HandleErrorEnv ()
-        constr2Map = product2Map . view #product
-
-        product2Map :: PC.Product -> HandleErrorEnv ()
-        product2Map = \case
-          PC.RecordI r -> record2Map r
-          PC.TupleI t -> tuple2Map t
-
-        record2Map :: PC.Record -> HandleErrorEnv ()
-        record2Map r = traverse_ field2Map $ r ^. #fields
-
-        field2Map :: PC.Field -> HandleErrorEnv ()
-        field2Map = ty2Map . view #fieldTy
-
-        ty2Map :: PC.Ty -> HandleErrorEnv ()
-        ty2Map = \case
-          PC.TyVarI tvar ->
-            tell @IETermMap $ M.singleton (tyVar2Variable tvar) (Left tvar)
-          PC.TyAppI tapp -> do
-            ty2Map $ tapp ^. #tyFunc
-            traverse_ ty2Map $ tapp ^. #tyArgs
-          PC.TyRefI tyref -> do
-            var <- tyRef2Variable tyref
-            tell @IETermMap $ M.singleton var (Right tyref)
-
-        tuple2Map :: PC.Tuple -> HandleErrorEnv ()
-        tuple2Map = traverse_ ty2Map . view #fields
+      I.InferUnboundTermErr uA -> do
+        case uA of
+          ForeignRef fr ->
```

Comment:
TODO: Rename this as it's a confusing misnomer, call it QTyRef and say `type QTyRef = (PC.ModuleName, PC.TyName` (and perhaps add the field with the 'original' PC.TyRef to aid in reporting).

'Foreign' only makes sense if you have 'Local'.